File: 3133/solution.py
class Solution:
    def minEnd(self, n: int, x: int) -> int:
        n -= 1
        i = 0
        while n > 0:
            bit = n & 1
            while (1 << i) & x:
                i += 1
            x |= bit << i
            i += 1
            n >>= 1
        return x

        # time log(n) + log(x)
        # space O(1)

File: 3133/solution.java
class Solution {
    public long minEnd(int n, int x) {
        int i = 0;
        long m = n;
        long res = x;
        m--;

        while (m > 0) {
            while (((1L << i) & res) != 0) {
                i++;
            }
            res |= ((m & 1) << i);
            i++;
            m >>= 1;
        }
        return res;
    }
}

// time O(1)
// space O(1)